Library/Homebrew/utils/io_selector.rb

```diff
@@ -0,0 +1,67 @@
+require "delegate"
+
+require "extend/io"
+
+module Utils
+  class IOSelector < DelegateClass(Hash)
+    attr_reader :separator
+
+    alias all_streams keys
+    alias all_tags values
+    alias tag_of fetch
+
+    def self.each_line_from(streams = {}, separator = $/, &b)
+      new(streams, separator).each_line_nonblock(&b)
+    end
+
+    def each_line_nonblock(&b)
+      each_line_nonblock_enum.each(&b)
+    end
+
+    def pending_streams
+      @pending_streams ||= all_streams.dup
+    end
+
+    def initialize(streams = {}, separator = $/)
+      super(streams.invert.compare_by_identity)
+      @separator = separator
+    end
+
+    private
+
+    def each_line_nonblock_enum
+      Enumerator.new do |yielder|
+        each_readable_stream_until_eof do |stream|
+          result = stream.readline_nonblock(separator) || ""
+          yielder << [tag_of(stream), result]
+        end
+        close_all
+      end
+    end
+
+    def each_readable_stream_until_eof(&b)
+      loop do
+        select_streams.each do |stream|
+          pending_streams.delete(stream) if stream.eof?
+          yield_gracefully(stream, &b)
+        end
+        break if pending_streams.empty?
+      end
+    end
+
+    def select_streams
+      IO.select(pending_streams)[0]
```

Comment:
Thanks for your feedback. While I technically agree, I beg to differ [from an engineering point of view](https://softwareengineering.stackexchange.com/questions/107669/one-line-functions-that-are-called-only-once), even more so as we’re dealing with Ruby code. As Avdi Grimm puts it in _Confident Ruby,_

> Writing maintainable code is all about focusing our mental energy on one task at a time, and equally importantly, on one level of abstraction at a time.

That maintainability thing is why I almost never inline such methods in Ruby.

Thoughts?